text/0000-glimmer-component-signature.md

```diff
@@ -0,0 +1,361 @@
+   - Can I pass attributes to it?
+   - If I apply a modifier, what kind(s) of DOM `Element` object will the modifier see?
+
+In the Glimmer `Component` base class today, a TypeScript component author can answer the first bucket of questions above by defining an `Args` type for the component class. Using this structured documentation, developer tooling can surface information about a component's arguments to consumers without requiring them to go read either ad-hoc prose or the implementation itself to determine the component's intended usage.
+
+For the second and third sets of questions, however, there is no such affordance for providing structured answers.
+
+The goal of this change is to enable tooling to provide a more complete picture for developers of the ways in which Glimmer components can be used. The word "tooling" here is used in broad terms, including:
+ - API/design system documentation generators, like [Storybook] or [Hokulea]
+ - Language servers that provide hover information, go-to-definition, etc. like [TypeScript] or [uELS]
+ - Template-aware static analysis and typechecking systems, like [Glint] or [`els-addon-typed-templates`]
+
+[Storybook]: https://storybook.js.org/
+[Hokulea]: https://github.com/hokulea/hokulea
+[TypeScript]: https://www.typescriptlang.org/
+[uELS]: https://github.com/lifeart/ember-language-server
+[Glint]: https://github.com/typed-ember/glint
+[`els-addon-typed-templates`]: https://github.com/lifeart/els-addon-typed-templates
+
+## Detailed design
+
+This RFC proposes that a component written like this today:
+
+```ts
+import Component from '@glimmer/component';
+
+export interface GreetingArgs {
+  /** The greeting to display. */
+  message: string;
+
+  /** Optionally, who to greet. */
+  target?: string;
+}
+
+export default class Greeting extends Component<GreetingArgs> {
+  // ...
+}
+```
+
+Would instead be written like this:
+
+```ts
+import Component from '@glimmer/component';
+
+export interface GreetingSignature {
+  Args: {
+    /** The greeting to display. */
+    message: string;
+
+    /** Optionally, who to greet. */
+    target?: string;
+  };
+}
+
+export default class Greeting extends Component<GreetingSignature> {
+  // ...
+}
+```
+
+By nesting the `Args` type under a top-level _signature_ type, we create a place to provide additional type information for other aspects of a component's behavior. One key element of this design is that eases future evolution of the signature, as including additional keys with new meaning won't be a breaking change.
+
+In addition to `Args`, two other signature members are proposed here: `BlockParams` and `Element`. A detailed explanation of each of these follows.
+
+### `Args`
+
+The `Args` signature member works exactly as the top-level `Args` type parameter did in the past. That is, it determines the type of `this.args` in the component backing class, as well as acting as an anchor for human-readable documentation for specific arguments.
```

Comment:
"did in the past" is a dead reference (what's the past from this point of the RFC?). Would move the example from the beginning of that broader section here, so it can be defined with "in the past" and makes it logically persistent within this article.

Reply:
As above I think this comes down to viewing the proposal not as the new state in a vacuum but as the _diff_ between the old state and the new one, but I'll see if I can rework the phrasing here to avoid confusion 🙂